### PR TITLE
add complex subtypes support

### PR DESCRIPTION
**Pull request by vintikzzz**
Just sequel of [1782](https://github.com/elixir-ecto/ecto/pull/1782)
[Cassandra DB](http://cassandra.apache.org/) supports different complex types such as [tuples](http://docs.datastax.com/en/cql/3.1/cql/cql_reference/tupleType.html), tuple of tuples, [maps of
maps](http://docs.datastax.com/en/cql/3.1/cql/cql_reference/tupleType.html), etc… And it is really important to modify a little bit type
validation. So migrations could look like:
```elixir
create table(@table, primary_key: false) do
  add :id, :uuid, primary_key: true
  add :value1, {:tuple, [:integer, :string, :float]}
  add :value2, {:tuple, [:integer, :string, {:tuple, [:integer, :integer]}]}
  add :value3, {:map, {:integer, :integer}}
end
```
For types with defined number of subtypes we use tuples (example `{:map, {:integer, :integer}}`)
For types with undefined number of subtypes we use lists (example `{:tuple, [:integer, :string, :float]}`)
If we want to make tuple with only one element we use just single atom (example `{:tuple, :integer}`

### CODE REVIEW

File: lib/ecto/migration.ex [hunk 1]
```diff
@@ -801,6 +801,14 @@ defmodule Ecto.Migration do
     validate_type!(subtype)
   end
 
+  defp validate_type!({type, subtype}) when is_atom(type) and is_list(subtype) do
+    for t <- subtype, do: validate_type!(t)
+  end
+
+  defp validate_type!({type, subtype}) when is_atom(type) and is_tuple(subtype) do
```
Comment (josevalim) [inline lib/ecto/migration.ex]: Do you need to support tuples with more than two elements? If not, I would prefer to strictly check for that. We also need unit tests for this.

Comment (vintikzzz) [inline lib/ecto/migration.ex]: Example migration for storing vector data requires 3 elements in a tuple:
```elixir
create table(:vector, primary_key: false) do
  add :id, :uuid, primary_key: true
  add :start_coords, {:tuple, [:integer, :integer, :integer]}
  add :end_coords, {:tuple, [:integer, :integer, :integer]}
end
```

Comment (josevalim) [inline lib/ecto/migration.ex]: That's a list though? And is that always 3 elements or it may be 2, 3, 4 and more?

Comment (vintikzzz) [inline lib/ecto/migration.ex]: Yes, it is like a list. But it has always 3 elements and every element must has proper type (in my example `integer`).

Comment (vintikzzz) [inline lib/ecto/migration.ex]: I think it might be a good option.

Comment (vintikzzz) [inline lib/ecto/migration.ex]: @josevalim any ideas?

Comment (josevalim) [inline lib/ecto/migration.ex]: @vintikzzz for all purposes it seems you don't have a use case for a list so I wouldn't support it until there is a need.

Comment (vintikzzz) [inline lib/ecto/migration.ex]: @josevalim so we can switch back to variant introduced in [1782](https://github.com/elixir-ecto/ecto/pull/1782). Am I right?

Comment (josevalim) [inline lib/ecto/migration.ex]: Yes, precisely. Sorry for the back and forth. :S

**Pull request closed without merging**